_posts/2019-06-27-list-strategy.md

```diff
@@ -0,0 +1,129 @@
+Example: the remote strategy silently falls back to sandboxed execution if an action can't run remotely. 
+The sandbox strategy silently falls back to non-sandboxed local execution if an action doesn't support sandboxing. 
+There's no good way to configure this behavior.
+
+**Example:** I want to configure my build to run remotely and fallback onto `local` strategy in case remote execution is not possible. My configuration would be something like:
+
+``` 
+$ bazel build
+ --spawn_strategy=remote
+ --strategy=Javac=remote
+ --strategy=Closure=remote
+ --strategy=Genrule=remote
+ --remote_local_fallback_strategy=local
+ ... 
+```
+
+
+## How it is after bazel 0.27
+- Bazel now *auto-detects* the execution strategy, if no strategy flag provided.  
+If none of the strategy flags was used, bazel will generate a default list of strategies `remote,worker,sandboxed,local` and, for every action it wants to execute, will pick up the first strategy that can execute it.
+
+- The user can pass comma-separated lists of strategies to the above mentioned flags: `--spawn_strategy=remote,worker,linux-sandbox`.  
+Each strategy now knows whether it can execute a given action.
+For any action that it wants to execute, Bazel just picks the first strategy from the given list (or from the default list) that claims to be able to execute the action. 
+
+- If action cannot be executed with any of the given strategies, build will fail.  
+If any of the strategy flags is provided, then bazel will use ONLY those strategies that are listed there. 
+Which means that `--spawn_strategy=remote` will NOT fallback to a `sandboxed` strategy or to any other strategy in case remote action failed or not possible at all.
+Moreover, `--remote_local_fallback_strategy=local` is now deprecated and even if it was provided, bazel will ignore it and build would still fail. 
+
+**For example**, if you try to execute `Genrule` rule remotely using `--spawn_strategy=remote`, you will get the following error:
+
+```
+ERROR: No usable spawn strategy found for spawn with mnemonic Genrule. Are your --spawn_strategy or --strategy flags too strict?
+```
+
+To fix it use a bazel *default* strategy config (remove strategy flags) or `--spawn_strategy=remote,local` or `--strategy=Gengule=local` and the build will be successful.
+
+### Why is it better?
+- Now you can control your execution strategy and have your build fail if any non-sandboxable or non-remotable action sneaks in! 
+More reproducibility and safety for your builds!
+- The strategies no longer do their own custom fallback, simplifying the code and unifying the behavior.
+- You might even completely forget about strategies configurations if the default behavior satisfies your needs.
+
+**Example:** This is how my first example will look like now:
+
+``` 
+$ bazel build
+ --spawn_strategy=remote,local
+ ... 
+```
+
+And if I want a custom strategy for my Java rules I can still add `--strategy=Javac=`:
```

Comment:
I'm a bit confused here. Maybe some more explanation?

In this example the "no-remote" Java rules would fall back to worker, but e.g., "no-remote" Scala rules would fall back to local.

I would say that a more common example would be using 
--spawn_strategy=remote,worker,local
Where anything "no-remote" that can fall back to worker would use worker, otherwise local (a configuration that AFAIK is not previously possible)

A more common example where a user might override per-mnemonic might be the other way: e.g., they don't want Javac actions to run remotely, only locally in the worker mode.
Then they'll have
 --spawn_strategy=remote,local
 --strategy=Javac=worker